content/backend/graphql-scala/7-relations.md

```diff
@@ -0,0 +1,457 @@
+
+```scala
+/**
+  * Load schema and populate sample data within this Sequence od DBActions
+  */
+val databaseSetup = DBIO.seq(
+  Users.schema.create,
+  Links.schema.create,
+  Votes.schema.create,
+
+  Users ++= Seq(
+    User(1, "mario", "mario@example.com", "s3cr3t"),
+    User(2, "Fred", "fred@flinstones.com", "wilmalove")
+  ),
+
+  Links ++= Seq(
+    Link(1, "http://howtographql.com", "Awesome community driven GraphQL tutorial",1, DateTime(2017,9,12)),
+    Link(2, "http://graphql.org", "Official GraphQL webpage",1, DateTime(2017,10,1)),
+    Link(3, "https://facebook.github.io/graphql/", "GraphQL specification",2, DateTime(2017,10,2))
+  ),
+
+  Votes ++= Seq(
+    Vote(1, 1, 1),
+    Vote(2, 1, 2),
+    Vote(3, 1, 3),
+    Vote(4, 2, 2),
+  )
+)
+
+```
+
+</Instruction>
+
+Because of relations, some operations should be executed in proper order.
+
+
+### Defining User->Link relation
+
+Lets begin from relation `User`-`Link`. In the first entity we have to add field `links` in the second `postedBy`, both defined by single relation.
+
+Because of relation with `User`, `Link` has two fields that could be used to fetch it by: `id` and `postedBy`.
+The first is default and used in normal way, the second is used if we want to find link for proper user.
+
+<Instruction>
+
+Add relation to able to find `Link`s by userId.
+
+```scala
+
+val linkByUserRel = Relation[Link, Int]("byUser", l => Seq(l.postedBy))
+```
+
+</Instruction>
+
+This kind Relation is of type `SimpleRelation` and has only two arguments: the first is name, and the second is a function which somehow extracts sequence of user id's from link entity. In our case is super easy, because `postedBy` has such id. All we need to do is wrap it into the sequence.
+
+Now we have to add this relation to the fetcher, to do this, we have to use `Fetcher.rel` function instead previously used `apply`
+
+<Instruction>
+
+Change `linksFetcher` to the following:
+
+```scala
+val linksFetcher = Fetcher.rel(
+    (ctx: MyContext, ids: Seq[Int]) => ctx.dao.getLinks(ids),
+    (ctx: MyContext, ids: RelationIds[Link]) => ctx.dao.getLinksByUserIds(ids(linkByUserRel))
+  )
+
+```
+
+</Instruction>
+
+What do we have here? As I mentioned above, now we're using `.rel` function. It needs the second function to be passed as the argument. This function is for fetching related data from datasource. In our case it uses a function `getLinksByUserIds` what we have to add to our dao.
+
+
+<Instruction>
+
+In `DAO` class add a function:
+
+```scala
+def getLinksByUserIds(ids: Seq[Int]): Future[Seq[Link]] = {
+    db.run {
+      Links.filter(_.postedBy inSet ids).result
+    }
+}
+```
+
+</Instruction>
+
+Actually we've simplified the code above a little. When you will thing more about the part `ctx.dao.getLinksByUserIds(ids(linkByUserRel)` you can ask "And what if link has two relations? Does `getLinkByUserIds` could be replaced by another function?" Be patient, such case will be covered later in this chapter.
```

Comment:
`When you will thing more` -> `When you look into ... a but, you can wonder ...`